public/app/features/templating/template_srv.ts

```diff
@@ -92,14 +92,21 @@ export class TemplateSrv implements BaseTemplateSrv {
     this.index[variable.name] = variable;
   }
 
-  getAdhocFilters(datasourceName: string) {
+  getAdhocFilters(datasourceName: string): AdHocVariableFilter[] {
     let filters: any = [];
+    let ds = getDataSourceSrv().getInstanceSettings(datasourceName);
+
+    if (!ds) {
+      return [];
+    }
 
     for (const variable of this.getAdHocVariables()) {
-      if (variable.datasource === null || variable.datasource === datasourceName) {
+      const variableUid = variable.datasource?.uid;
+
+      if (variableUid === ds.uid || (variable.datasource == null && ds?.isDefault)) {
         filters = filters.concat(variable.filters);
-      } else if (variable.datasource.indexOf('$') === 0) {
-        if (this.replace(variable.datasource) === datasourceName) {
+      } else if (variableUid?.indexOf('$') === 0) {
+        if (this.replace(variableUid) === datasourceName) {
           filters = filters.concat(variable.filters);
         }
       }
```

Comment:
This function was a bit tricky, as it's part something many data sources use (internal & external) to support ad hoc queries. I opted to just lookup the data source by name to get the uid to find the matching adhoc variable. 

We could also change this signature to `getAdhocFilters(dsUid: string)` as long as we always lookup the ds via getInstanceSettings incase external plugins call this function with the data source name.